### PR TITLE
Precompile for Poseidon2 BabyBear

### PR DESCRIPTION
**Pull request by lvella**

### CODE REVIEW

File: std/machines/hash/poseidon2_bb.asm [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    array::zip(x7, array::zip(x3, a, |x3, a| x3 * x3 * a), |x7, expected| x7 = expected);
+
+    // Apply S-Boxes on the first element and otherwise if it is a full round.
+    let b: expr[] = array::new(STATE_SIZE, |i| if i == 0 {
+        x7[i]
+    } else {
+        PARTIAL * (a[i] - x7[i]) + x7[i]
+    });
+
+    // The MDS matrix
```
Comment (chriseth) [inline std/machines/hash/poseidon2_bb.asm:166]: Can this be shared across the different poseidon implementations?

File: std/machines/hash/poseidon2_bb.asm [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    // Like LASTBLOCK, but also 1 in the last row of the table
+    // Specified this way because we can't access the degree in the match statement
+    pol constant LAST = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]* + [1];
+
+    // Whether the current round is a partial round
+    pol constant PARTIAL = [0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0]*;
+    
+    // The round constants
+    pol constant C_0 = [0xb585f766f2144405, 0x86287821f722c881, 0xe9fa634a21de0082, 0x92a756e67e2b9413, 0x3cc3f892184df408, 0x7131aa45268d7d8c, 0x99ad1aab0814283b, 0xeb84f608da56ef48, 0x7159cd30c3ac118e, 0xdcef0797c2b69ec7, 0xd0762cbc8ca6570c, 0x30a4680593258387, 0x15a16a8a8322d458, 0x5a3f1bb1c53a9645, 0x775005982d74d7f7, 0xf9cc95c22b4c1fcc, 0xc49366bb25e8513, 0xdd611f1000c17442, 0x2ff876fa5ef97c4, 0x3d06c8bd1514e2d9, 0xe89cd854d5d01d33, 0xece5a71e0cfedc75, 0x90004c1371b893c5, 0xde122bebe9a39368, 0x4d61e56a525d225a, 0x1478d361dbbf9fac, 0x475cd3205a3bdcde, 0xe70201e960cb78b8, 0x7be5b9ffda905e1c, 0xf3c12fe54d5c653b, 0x0]*;
```
Comment (chriseth) [inline std/machines/hash/poseidon2_bb.asm:128]: Can these be shared across the different poseidon implementations?

Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:128]: No, because the number of rounds depends on the state length, which depends on the field size. Also these constants don't fit in BabyBear.

File: std/machines/hash/poseidon2_bb.asm (hunk 1 above)
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:166]: No, because the power you use for the S-Box depends on the field. But 7 is used both by BabyBear and Goldilocks.

File: std/machines/hash/poseidon2_bb.asm [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    // Half the number of external rounds (half of external rounds happen before and half after the internal rounds).
+    let HALF_EXTERNAL_ROUNDS: int = 4;
+
+    // Number of internal rounds
+    let INTERNAL_ROUNDS: int = 13;
+
+    // External round MDS matrix
+    let MDS = [
+        [4, 6, 2, 2, 2, 3, 1, 1, 2, 3, 1, 1, 2, 3, 1, 1],
```
Comment (chriseth) [inline std/machines/hash/poseidon2_bb.asm:56]: Could we please avoid copying these magic constants across 4-5 files?

Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:56]: They are not. The only other place most of these constants appear is inside plonky3 rust code.

The best I can do is, after merging this branch with the poseidon_bb one, place a fraction of the big list of random numbers that is common between the two of them in a separate file, as they are the output of the same deterministic RNG.

File: std/machines/hash/poseidon2_bb.asm [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+    // Multiply with MDS Matrix
+    let dot_product = |v1, v2| array::sum(array::zip(v1, v2, |v1_i, v2_i| v1_i * v2_i));
+    let c = array::map(M, |M_row_i| dot_product(M_row_i, b));
+
+    // Copy c to state in the next row
+    array::zip(state, c, |state, c| (state' - c) * (1-LAST) = 0);
+
+    // In the last row, the first OUTPUT_SIZE elements of the state should equal output
+    array::zip(output, state, |output, state| LASTBLOCK * (output - state) = 0);
```
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm]: Done.

**Comment by lvella** [comment]
Blocked on #1933.

File: std/machines/hash/poseidon2_bb.asm [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+use std::machines::small_field::pointer_arith::increment_ptr;
+use std::machines::split::split_bb::SplitBB;
+
+// Implements the Poseidon2 permutation for BabyBear field.
+//
+// Apparently it can be used to hash arbitrary sized data by using the
+// Merkle–Damgård construction, or it can be used as a compression function
+// for building a Merkle tree.
+//
+// As it stands, it can not be used in a Sponge construction, because we don't
```
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:18]:
```suggestion
// As it stands, it cannot be used in a Sponge construction, because we don't
```

File: std/machines/hash/poseidon2_bb.asm [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    // https://github.com/Plonky3/Plonky3/blob/2df15fd05e2181b31b39525361aef0213fc76144/poseidon2/src/round_numbers.rs#L42
+
+    // S-box degree (this constant is actually not used, because we have to break the exponentiation into steps of at most degree 3).
+    let SBOX_DEGREE: int = 7;
+
+    // Number of field elements in the state
+    let STATE_SIZE: int = 16;
+
+    // Number of output elements
+    // (TODO: to use the Sponge construction, the entire state should be output)
```
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:58]: can you make an issue for this?

File: std/machines/hash/poseidon2_bb.asm [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    };
+
+    // Calculate the addresses and load all the inputs into the first time step
+    let input_addr_high: col[STATE_SIZE];
+    let input_addr_low: col[STATE_SIZE];
+    address_inc(input_addr_high, input_addr_low);
+
+    let input_low: col[STATE_SIZE];
+    let input_high: col[STATE_SIZE];
+    // TODO: when link is availailable inside functions, we can turn this into array operations.
```
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:197]:
```suggestion
    // TODO: when link is available inside functions, we can turn this into array operations.
```

File: std/machines/hash/poseidon2_bb.asm [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+    link if is_used ~> (input_high[0], input_low[0]) = mem.mload(input_addr_high[0], input_addr_low[0], time_step);
```
Comment (leonardoalt) [inline std/machines/hash/poseidon2_bb.asm:198]: @georgwiese all these links :grimacing:

File: std/machines/hash/poseidon2_bb.asm (hunk 6 above)
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:58]: I am not even sure we want it.

File: std/machines/hash/poseidon2_bb.asm [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+                    increment_ptr(4, high, low, next_high, next_low)
+                }
+            ), [],
+            |a, b| a + b
+        )
+    };
+
+    // Calculate the addresses and load all the inputs into the first time step
+    let input_addr_high: col[STATE_SIZE];
+    let input_addr_low: col[STATE_SIZE];
```
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:192]: I think this is for another time, but I wonder if there is a way not to commit to these values. Couldn't `increment_ptr ` return expressions for the incremented address instead of constraining an witness column to be equal to it? Not sure what happens when you pass that into `increment_ptr` again though, maybe you'd get a high-degree expression?

File: std/machines/hash/poseidon2_bb.asm [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    //
+    // Similarly, the output data is written to memory at the provided pointer.
+    //
+    // Reads happen at the provided time step; writes happen at the next time step.
+    operation poseidon2_permutation<0>
+        input_addr_high[0], input_addr_low[0],
+        output_addr_high[0], output_addr_low[0],
+        time_step ->;
+
+    let latch: col =  |_| 1;
```
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:41]:
```suggestion
    let latch = 1;
```
This works too and does not create a fixed column!

File: std/machines/hash/poseidon2_bb.asm (hunk 8 above)
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:198]: Don't be afraid of links ;)

In the optimized case, this should create 2 columns + some constraints for each link. But the whole machine has block size 1, so it will still be few cells to commit to. I think the main disadvantage of this approach vs larger block size & fewer links is increased verifier time (because it scales with the number of columns & constraints). We'll have to see what the tradeoff space is here.

File: std/machines/hash/poseidon2_bb.asm [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+                external_round(round_idx, pre_state, post_state);
+                post_state
+            }
+        );
+
+        // Perform the internal rounds
+        let after_internal_rounds = utils::fold(
+            INTERNAL_ROUNDS, |round_idx| round_idx, after_initial_rounds,
+            constr |pre_state, round_idx| {
+                let post_state = array::new(STATE_SIZE, |_| { let x; x});
```
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:237]: Again, we should optimize this later, but do you think we have to commit to all the intermediate states? Could `{internal,external}_round()` return an expression and we just use that?

Since Poseidon is an algebraic hash function, I think we could in principle express the entire function as one big algebraic expression. The only problem would be that its would be of too high degree. But eventually we should try to minimize the witness columns.

**Review by georgwiese — COMMENTED** [review]
This looks nice! :) I have a few questions, but I think many of my suggestions could be implemented in future PRs.

File: plonky3/src/bin/gen_poseidon2_bb_consts.rs [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+        *baby_bear::ROUNDS_P,
+        poseidon2::internal_constants::<BabyBear>(*ROUNDS_P),
+        p3_baby_bear::DiffusionMatrixBabyBear::default(),
+    );
+
+    let test_vectors = [
+        [BabyBear::zero(); WIDTH],
+        [BabyBear::one(); WIDTH],
+        [-BabyBear::one(); WIDTH],
+        // The test vector from goldilocks, that I don't know where it came from.
```
[the remark this answers is not in the document]
Comment (lvella) [reply plonky3/src/bin/gen_poseidon2_bb_consts.rs]: fixed

File: std/machines/hash/poseidon2_bb.asm [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+use std::convert::expr;
+use std::machines::small_field::memory::Memory;
+use std::machines::small_field::pointer_arith::increment_ptr;
+use std::machines::split::split_bb::SplitBB;
+
+// Implements the Poseidon2 permutation for BabyBear field.
+//
+// Apparently it can be used to hash arbitrary sized data by using the
+// Merkle–Damgård construction, or it can be used as a compression function
+// for building a Merkle tree.
```
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:16]: > Hmm, so the function permutes a 512-Bit input and returns the first 256 Bit, right?

Right. This is the version used by p3 backend, so it should be useful for recursion.

> This is different from the GL Poseidon, which permutes a 768-Bit input (12 64-Bit field elements) and returns the first 256 Bit. There, you'd input your 512-Bit input (e.g. the two child hashes in a Merkle tree) plus a 256-Bit capacity, which would be a constant (where different constants give you different hash functions).

Yes.

> Are you sure we're not supposed to put in a 256-Bit constant here as well? How would we get different hash functions (e.g. one for the Merkle leaves and one for the internal nodes)?

For internal nodes a full hash function is not necessary, just a compression is sufficient. Which this is.

For the leaves, I don't know for sure. How do you do with GL?

I can see 2 options. Alternative a) is Merkle-tree all the way down to 512 bits of data (it works because the data is a fixed power of 2 size). I guess this is safe because this is an intended usage in Poseidon2 paper.

Alternative b) is to use a construction that can turn a compression function into a hash function, like Merkle–Damgård (or its updated version, HAIFA construct). This usage is just implied in passing in Poseidon2 paper. If you need 2^256 different hash functions, set the IV to something other than zero in the first compression step, but I don't see why it would be needed in the bootloader.

Alternative a) seems simpler to me.

File: std/machines/hash/poseidon2_bb.asm [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+        let x3 = array::map(step_a, constr |x| { let x3; x3 = x * x * x; x3});
+        let x7 = array::zip(x3, step_a, constr |x_3, x| { let x7; x7 = x_3 * x_3 * x; x7 });
+
+        // Multiply with MDS Matrix
+        array::zip(output, apply_mds(x7, array::len(output)), |out, x| out = x);
+    };
+
+    let internal_round = constr |c_idx, input, output| {
+        // Add constant (weird, I thought the entire state was used here,
+        // but this is how Plonky3 does it).
```
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:147]: Notice how the "AddRoundConstants(·)" is still applied to the entire state vector in this figure, not only the s-boxed element. In the p3 version, a constant is only added to state[0]. I am not sure how this affects security.

File: std/machines/hash/poseidon2_bb.asm [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+        output[0] = (line_sum + DIFF_DIAGONAL[0] * x7) * DIFF_MULTIPLIER;
+        array::zip(
+            array::zip(
+                array::sub_array(input, 1, STATE_SIZE - 1),
+                array::sub_array(output, 1, STATE_SIZE - 1),
+                constr |in_v, out_v| (in_v, out_v)
+            ),
+            array::sub_array(DIFF_DIAGONAL, 1, STATE_SIZE - 1),
+            constr |(in_v, out_v), diag| out_v = (line_sum + diag * in_v) * DIFF_MULTIPLIER
+        );
```
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm]: Added comment explaining the matrix and how the multiplication works.

The whole point of Poseidon2 is that this multiplication can be heavily optimized.

File: std/machines/hash/poseidon2_bb.asm (hunk 9 above)
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:218]: I am pretty sure it would extrapolate the degree limit, due to the logic to juggle the carry between the two limbs.

File: std/machines/hash/poseidon2_bb.asm (hunk 11 above)
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:263]: Yes.

See the issue I opened related to this: #1971

File: std/machines/hash/poseidon2_bb.asm [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+        array::fold(
+            array::zip(
+                array::sub_array(addr, 0, array::len(addr) - 1),
+                array::sub_array(addr, 1, array::len(addr) - 1),
+                constr |(high, low), (next_high, next_low)| {
+                    increment_ptr(4, high, low, next_high, next_low)
+                }
+            ), [],
+            |a, b| a + b
+        )
```
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:217]: I believe this was my first attempt. Let me try again.

[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:217]: Nope.

```
Cannot unify types () and std::prelude::Constr
    ┌─ /home/lucas/powdr/src/powdr/std/machines/hash/poseidon2_bb.asm:208:13
    │  
208 │ ╭             array::zip(
209 │ │                 array::sub_array(addr, 0, array::len(addr) - 1),
210 │ │                 array::sub_array(addr, 1, array::len(addr) - 1),
211 │ │                 constr |(high, low), (next_high, next_low)| {
212 │ │                     increment_ptr(4, high, low, next_high, next_low);
213 │ │                 }
214 │ │             );
    │ ╰─────────────^

thread 'poseidon2_bb_test' panicked at /home/lucas/powdr/src/powdr/pipeline/src/test_util.rs:325:51:
called `Result::unwrap()` on an `Err` value: ["Expected type std::prelude::Constr[] but got type ()[].\nCannot unify types () and std::prelude::Constr at /home/lucas/powdr/src/powdr/std/machines/hash/poseidon2_bb.asm:9681-9986"]
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: std/machines/hash/poseidon2_bb.asm [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+        array::map(array::sub_array(MDS, 0, output_len), |row| dot_product(row, input))
+    };
+
+    let external_round = constr |c_idx, input, output| {
+        // Add constants
+        let step_a = array::zip(input, EXTERNAL_ROUND_CONSTANTS[c_idx], |v, c| v + c);
+
+        // Apply S-box
+        let x3 = array::map(step_a, constr |x| { let x3; x3 = x * x * x; x3});
+        let x7 = array::zip(x3, step_a, constr |x_3, x| { let x7; x7 = x_3 * x_3 * x; x7 });
```
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm]: Done.

File: std/machines/hash/poseidon2_bb.asm (hunk 11 above)
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:267]: Actually, not really. If you multiply 2 polynomials, the resulting polynomial has size m*n, so they can grow to intractable sizes.

Unless we internally represent polynomials as a DAG of multiplication of polynomial factors (which I guess neither we nor the backends do).

File: std/machines/hash/poseidon2_bb.asm (hunk 13 above)
[the remark this answers is not in the document]
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:16]: > For internal nodes a full hash function is not necessary, just a compression is sufficient. Which this is.

What do you mean by "compression is sufficient"? I think we need collision resistance, right? I would *guess* that this construction would be collision resistant, I'm just wondering if that's a standard implementation or whether we do something unusual here.

And yes, for the bootloader the Merkle tree in the bootloader has a pre-defined depth, so it is fine to use the same hash function!

File: std/machines/hash/poseidon2_bb.asm (hunk 11 above)
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:267]: > If you multiply 2 polynomials, the resulting polynomial has size m*n, so they can grow to intractable sizes.

I don't understand this comment! If you multiply two polynomials of degrees m and n, you get a polynomial of degree m + n, right?

I do think that #1971 would fix this: We'd add each intermediate state as an intermediate polynomials, which depending on the degree bound would be inlined or materialized as a witness column.

File: std/machines/hash/poseidon2_bb.asm [hunk 18]
```diff
@@ -148,12 +155,35 @@ machine Poseidon2BB(mem: Memory, split_BB: SplitBB) with
...
+        //                   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, O, 1, 1]
+        //                   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, P, 1]
+        //                   [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, Q]
+        //
+        // Where A, B, C, ..., Q are the elements of the DIFF_DIAGONAL array plus 1.
+        //
+        // The idea of using such matrix in Poseidon2 is that, instead of performing
+        // a full matrix multiplication, we can optimize it by summing the elements
+        // of the input vector, and then adjusting each output[k] element by
+        // input[k] * DIFF_DIAGONAL[k].
```
Comment (georgwiese) [inline std/machines/hash/poseidon2_bb.asm:186]: ❤️

File: std/machines/hash/poseidon2_bb.asm (hunk 14 above)
[the remark this answers is not in the document]
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:154]: Right! Strange indeed, but I guess we want to by consistent with Plonky3.

**Review by georgwiese — APPROVED** [review]
Cool!

File: std/machines/hash/poseidon2_bb.asm (hunk 11 above)
Comment (georgwiese) [reply std/machines/hash/poseidon2_bb.asm:267]: Oh, maybe you're talking about the fact that if e.g. if you repeatedly compute an S-Box $x^7$ and $x$ is an expression and you inline everything, the resulting expression grows exponentially? That's fixed by intermediate columns and Thibaut just pushed #1995, which makes the Plonky3 backend aware of them :)

File: std/machines/hash/poseidon2_bb.asm (hunk 13 above)
[the remark this answers is not in the document]
Comment (lvella) [reply std/machines/hash/poseidon2_bb.asm:16]: This is a quote from Poseidon2 paper that I interpret as meaning it is safe to use a single Poseidon2 permutation to level-up a Merkle-tree:

> While hash functions are perfectly usable and allow to make strong security
arguments, they are often more generic. For many classical applications (e.g.,
ordinary hashing of arbitrary-length inputs) this is not a major issue. However,
when building a Merkle tree with small fixed-size input lengths, we often only
need a single permutation call to process the entire input and compute the
desired output. In this case, the construction is not used as a hash function in a
traditional sense, but rather as a compression function.
